bitcoin/src/address.rs

```diff
@@ -499,7 +499,7 @@ impl Payload {
     pub fn p2wsh(script: &script::Script) -> Payload {
         let prog = WitnessProgram::new(
             WitnessVersion::V0,
-            script.wscript_hash().as_ref().to_vec()
+            script.wscript_hash().as_inner().to_vec()
```

Comment:
Oh we could remove an allocation by using `into_inner().to_vec()`, right?

Reply:
Is my assumption correct: `to_vec` just uses the array and doesn't reallocate but for a slice it has to allocate?

Reply:
No, `to_vec` will produce a vector, it won't output an array under any circumstances. (And a `Vec` cannot use an array as a backing store.)

Reply:
Oh so a `to_vec` call always does a heap allocation even if the vector is only used in a single scope. Interesting, I always assumed otherwise. Cheers.

Reply:
So you are saying that in the following case `to_vec` does a heap allocation.
```rust
{
    let a = [0_u8; 32]
    let v = a.to_vec();
    // do something with v
} // vec dropped here, so `a` was on the stack and v used a heap allocation
```
I suppose I should just look in the stdlib instead of being lazy and asking, I'll do that now.

Reply:
Yep you are correct, for anyone else reading `to_vec` seems to come from `slice::ConvertVec` and is implemented for `&[T]` where `T: Clone`. During the `to_vec` method call the `&` is added and `&[0_u8; 32]` is used as the receiver, the array is iterated and each byte is copied using `clone`. Cool, TIL.

Reply:
In general, Rust doesn't do crazy-allocation-removing optimizations. If you create non-empty `Vec`, `Box`, `Rc`, `Arc`... you're practically guaranteed to get a heap allocation.

I also recently noticed that having arrays rather than vec *even in test code* can help refactoring and such.